src/dist/matrix_ops.cpp

```diff
@@ -25,55 +23,89 @@ void rectangle_block(const Eigen::VectorXf &longDists,
-  std::vector<long> j_vec;
-  if (distCutoff > 0) {
-    for (long i = 0; i < denseDists.rows(); i++) {
-      for (long j = i + 1; j < denseDists.cols(); j++) {
-        if (denseDists(i, j) < distCutoff) {
-          dists.push_back(denseDists(i, j));
-          i_vec.push_back(i);
-          j_vec.push_back(j);
-        }
-      }
+sparse_coo sparsify_dists(const NumpyMatrix &denseDists,
+                          const float distCutoff,
+                          const unsigned long int kNN,
+                          const size_t num_threads)
+{
+    if (kNN > 0 && distCutoff > 0)
+    {
+        throw std::runtime_error("Specify only one of kNN or distCutoff");
+    }
+    else if (kNN < 1 && distCutoff < 0)
+    {
+        throw std::runtime_error("kNN must be > 1 or distCutoff > 0");
     }
-  } else if (kNN >= 1) {
-    // Only add the k nearest (unique) neighbours
-    // May be >k if repeats, often zeros
-    for (long i = 0; i < denseDists.rows(); i++) {
-      unsigned long unique_neighbors = 0;
-      float prev_value = -1;
-      for (auto j : sort_indexes(denseDists.row(i))) {
-        if (j == i) {
-          continue; // Ignore diagonal which will always be one of the closest
+
+    // Parallelisation parameters
+    const size_t num_samples = denseDists.rows();
+    size_t len = 0;
+  
+    // ijv vectors
+    std::vector<std::vector<float>> dists(num_samples);
+    std::vector<std::vector<long>> i_vec(num_samples);
+    std::vector<std::vector<long>> j_vec(num_samples);
+    if (distCutoff > 0)
+    {
+#pragma omp parallel for schedule(static) num_threads(num_threads) reduction(+:len)
+        for (long i = 0; i < denseDists.rows(); i++)
+        {
+            for (long j = i + 1; j < denseDists.cols(); j++)
+            {
+                if (denseDists(i, j) < distCutoff)
+                {
+                    dists[i].push_back(denseDists(i, j));
+                    i_vec[i].push_back(i);
+                    j_vec[i].push_back(j);
+                }
+            }
+            len += i_vec[i].size();
         }
-        bool new_val = abs(denseDists(i, j) - prev_value) < epsilon;
-        if (unique_neighbors < kNN || new_val) {
-          dists.push_back(denseDists(i, j));
-          i_vec.push_back(i);
-          j_vec.push_back(j);
-          if (!new_val) {
-            unique_neighbors++;
-            prev_value = denseDists(i, j);
-          }
-        } else {
-          break;
+    }
+    else if (kNN >= 1)
+    {
+        // Only add the k nearest neighbours
+#pragma omp parallel for schedule(static) num_threads(num_threads) reduction(+:len)
+        for (long i = 0; i < num_samples; ++i)
+        {
+            unsigned long counter = 0;
+            for (auto j : sort_indexes(denseDists.row(i)))
+            {
+                if (j == i)
+                {
+                    continue; // Ignore diagonal which will always be one of the closest
+                }
+                if (counter < kNN)
```

Comment:
I think you want the `new_val` check, as in the old code? Otherwise this always adds exactly `k`, and you can use the more easily parallelisable version

Reply:
Deleted